### PR TITLE
Only allocate partial shards to nodes with cache

### PR DESCRIPTION
**Pull request by DaveCTurner**
This commit introduces the `HasFrozenCacheAllocationDecider` which
restricts allocation of partial shards only to those nodes with a
configured shared cache, and also the `FrozenCacheInfoService` which
tracks which nodes do/don't have a frozen cache configured.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            this.rerouteService = rerouteService;
+            this.discoveryNode = discoveryNode;
+            this.nodeStateHolder = nodeStateHolder;
+        }
+
+        @Override
+        protected void doRun() {
+            client.admin()
+                .cluster()
+                .prepareNodesInfo(discoveryNode.getId())
```
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java:134]: I'm wondering if this works with security (see #67416). As said before, I have a slight preference for a dedicated transport action.

**Review by ywelsch — COMMENTED** [review]
I've left some comments, looking good overall.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java [hunk 2]
```diff
@@ -60,8 +60,8 @@ protected Settings nodeSettings(int nodeOrdinal) {
                 CacheService.SNAPSHOT_CACHE_SIZE_SETTING.getKey(),
                 rarely()
                     ? randomBoolean()
-                        ? new ByteSizeValue(randomIntBetween(0, 10), ByteSizeUnit.KB)
-                        : new ByteSizeValue(randomIntBetween(0, 1000), ByteSizeUnit.BYTES)
+                        ? new ByteSizeValue(randomIntBetween(1, 10), ByteSizeUnit.KB)
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java]: Ah oops I was using the wrong `SNAPSHOT_CACHE_SIZE_SETTING` to start with, failed to revert this. Fixed in 43959221eeee027b227d7d0228e16c26292810cf.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+    public void initialize(ClusterService clusterService) {
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        isElectedMaster = event.localNodeMaster();
+        if (isElectedMaster == false) {
+            clear();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java]: Ok, we were using `cleanCaches()` too so I removed this, see 70d398bea2b.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java:110]: Ok, I'll work on an internal transport action for this next.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java [hunk 4]
```diff
@@ -165,6 +191,11 @@ private AllocateUnassignedDecision decideAllocation(RoutingAllocation allocation
             return AllocateUnassignedDecision.no(UnassignedInfo.AllocationStatus.FETCHING_SHARD_DATA, null);
         }
 
+        if (SNAPSHOT_PARTIAL_SETTING.get(allocation.metadata().index(shardRouting.index()).getSettings())
+            && frozenCacheInfoService.isFetching()) {
+            return AllocateUnassignedDecision.no(UnassignedInfo.AllocationStatus.FETCHING_SHARD_DATA, null);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotAllocator.java:196]: Yes: if all the nodes are either a `NO` or still fetching (hence `THROTTLE`) then we exit on the next branch before even calling `fetchData`, so allocation explain returns `DECIDERS_THROTTLED`, whereas once one node is a `YES` then the response becomes `FETCHING_SHARD_DATA`.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/FrozenCacheInfoService.java]: Done in https://github.com/elastic/elasticsearch/pull/69904/commits/71b5bbb914b4fa644d1900e9f6584f290ba90739 now.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**